p2p/socket.go

```diff
@@ -0,0 +1,461 @@
+	closed  chan uint // this channel signals that the socket has been closed by receiving a new input
+
+	errored chan uint // on error, this channel receives a new input to signal the happening of an error
+	err     struct {  // the reference to store the encountered error.
+		sync.Mutex
+		error
+	}
+
+	// TODO(team): replace with whatever logged decided upon when telemetry hits the code base
+	logger struct { // a temporary logger struct to allow flexible injection of log functions
+		sync.Mutex
+		print func(...interface{}) (int, error)
+	}
+}
+
+// retrieves the network connection in question through the connector argument
+// and starts the IO operations on that connection, while putting in place event handlers for onSocketOpened and onSocketClosed events
+// returns an error on failure
+func (s *socket) open(ctx context.Context, connector func() net.Conn, onopened SocketEventMonitor, onclosed SocketEventMonitor) error {
+	s.buffers.write.Open()
+
+	conn := connector()
+
+	addr := ctx.Value("address").(string)
+	kind := ctx.Value("kind").(SocketKind)
+
+	if addr == "" {
+		return ErrSocketEmptyContextValue("address")
+	}
+
+	if kind == "" {
+		return ErrSocketEmptyContextValue("kind")
+	}
+
+	switch kind {
+	case Outbound:
+	case Inbound:
+	default:
+		s.close()
+		return ErrSocketUndefinedKind(string(kind))
+	}
+
+	go s.startIO(ctx, kind, addr, conn, onopened, onclosed)
+
+	select {
+	case <-s.errored:
+		return s.err.error
+	case <-s.reading:
+	}
+
+	select {
+	case <-s.errored:
+		return s.err.error
+	case <-s.writing:
+	}
+
+	s.isOpen.Store(true)
+	close(s.ready)
+
+	return nil
+}
+
+// Stops the ongoing IO operations gracefully
+// closes the network connection and closes all opened channels and signals that the close has went gracefully
+// does not block
+func (s *socket) close() {
+	if !s.isOpen.Load() {
+		return
+	}
+	close(s.done)
+
+	s.err.Lock()
+	close(s.errored)
+	s.err.Unlock()
+
+	s.buffers.write.Close()
+
+	s.isOpen.Store(false)
+
+	if s.conn != nil {
+		s.conn.Close()
+	}
+
+	s.closed <- 1
+	s.isOpen.Store(false)
+}
+
+// creates a reader and writer for the network connection
+// kicks off the onSocketOpened event handler, returns if there is an error
+// launches both the reading and writing routines (read, write methods): both are blocking
+// when the write routine exists, the onSocketClosed event handler kicks off (returns an error if there is one)
+// the write and read routines will both exist for the same reasons, so having just one of them block is sufficient.
+func (s *socket) startIO(ctx context.Context, kind SocketKind, addr string, conn net.Conn, onopened SocketEventMonitor, onclosed SocketEventMonitor) {
+	defer func() {
+		s.close()
+	}()
+
+	s.addr = addr
+	s.kind = kind
+
+	s.conn = conn
+	s.reader = bufio.NewReader(conn)
+	s.writer = bufio.NewWriter(conn)
+
+	if err := onopened(ctx, s); err != nil {
```

Comment:
Not blocking, just a comment:
I'm not a fan of placeholder code. Looking through the tests, you're just using these callbacks for telemetry tracking. I would prefer if the telemetry engine was built into the lifecycle with a more linear design

Reply:
🤔 deserves some thought, you are correct in assuming that these callbacks served primarily with tests, and their importance will be more apparent in the incoming PRs since they will be used by other parts of the p2p module.

I would say that I can benefit from a discussion about the linear design approach. Adding this to the backlog of topics.

Reply:
Added as a backlog item for protocol hours, resolving for now.